game-core/src/test/java/games/strategy/triplea/odds/calculator/DummyPlayerTest.java

```diff
@@ -0,0 +1,189 @@
+package games.strategy.triplea.odds.calculator;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.PlayerId;
+import games.strategy.engine.data.ResourceList;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.data.CasualtyDetails;
+import games.strategy.triplea.delegate.data.CasualtyList;
+import java.util.Collections;
+import java.util.List;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class DummyPlayerTest {
+
+  private final List<Unit> unitPool =
+      List.of(
+          mock(Unit.class), mock(Unit.class), mock(Unit.class), mock(Unit.class), mock(Unit.class));
+
+  private final List<Unit> orderOfLosses =
+      List.of(unitPool.get(2), unitPool.get(0), unitPool.get(1), unitPool.get(4), unitPool.get(3));
+
+  private final List<Unit> damaged = List.of(unitPool.get(0), unitPool.get(2));
+  private final List<Unit> killed = List.of(unitPool.get(1), unitPool.get(0));
+
+  private final CasualtyList defaultCasualties = mock(CasualtyList.class);
+
+  @BeforeEach
+  void setUp() {
+    when(defaultCasualties.getDamaged()).thenReturn(damaged);
+    when(defaultCasualties.getKilled()).thenReturn(killed);
+  }
+
+  @Test
+  void testSelectCasualties_oneLand_noLossOrder() {
+    for (final var unit : unitPool) {
+      final var unitType = mock(UnitType.class);
+      when(unit.getType()).thenReturn(unitType);
+      final var unitAttachment = mock(UnitAttachment.class);
+      when(unitType.getAttachment(any())).thenReturn(unitAttachment);
+      when(unitAttachment.getIsAir()).thenReturn(!unit.equals(unitPool.get(0)));
+      final var gameData = mock(GameData.class);
+      when(unit.getData()).thenReturn(gameData);
+      when(gameData.getResourceList()).thenReturn(mock(ResourceList.class));
+      final var playerId = mock(PlayerId.class);
+      when(unit.getOwner()).thenReturn(playerId);
+    }
+
+    final DummyPlayer player = new DummyPlayer(null, false, "", null, true, 0, 0, false);
+    final CasualtyDetails details =
+        player.selectCasualties(
+            unitPool,
+            null,
+            0,
+            null,
+            null,
+            null,
+            null,
+            null,
+            false,
+            null,
+            defaultCasualties,
+            null,
+            null,
+            false);
+    assertThat(details.getDamaged(), is(damaged));
```

Comment:
I'm seeing a consistent three assertions, and only one looks to be changing. I wonder if the tests could be more focused. For one aspect, given the first and third assertions are always the same, it could be that the method under test happens to produce the right results for this case but not in all cases. I'm pretty convinced the 'getKilled' is relatively well exercised.

I wonder then if we can make the tests 'more' focused on specific behaviors by adding some nesting. The 3 existing tests would be a single nested class for the `details.getKilled` check. We'd then add two additional nesting to test specifically the `getAutoCalculated` and `getDamaged`. It's also a question whether the full setup is needed to test those two methods.
What are your thoughts @RoiEXLab ?

Reply:
I tried to reduce the boilerplate a little bit. Tell me what you think

Reply:
Looking good. Boiler plate in test is not necessarily bad; though there is a trade-off maintenance cost of updating many things when things change. So a bit of a balance I suppose.

To be direct, what's the most painful in tests is when they fail and it's not clear what they were testing or should have passed to begin with. I'm a big fan of assertion messages describing conditions, things like `assertThat("this should pass because..", a, is(b));`

I also really like javadocs on test methods that describe the strategy being employed to test.

Overall it's really easy know why tests should pass when writing them, and why each case is necessary, but then over time that context is lost.

Hence, anything that can be done to help carry forward the information about why the test cases are expected to pass, what they are testing, could be of big value. On the flip side, having many similar tests to update and them failing, and not being sure why they even should have passed in the first place or what they were testing, can become a very large maintenance cost.